src/server/session.ts

```diff
@@ -2015,4 +2018,40 @@ namespace ts.server {
             }
         }
     }
+
+    /* @internal */ // exported for tests only
+    /** Since we will perform the rename in a changed document, line and offset information must be adjusted if the new text adds/deletes newlines. */
+    export function fixupLineAndOffset(oldPosition: protocol.Location, oldText: string, location: number, locationFileName: string, edits: ReadonlyArray<FileTextChanges>): protocol.Location {
+        let { line, offset } = oldPosition;
+        const lineStartOfLocation = location - (offset - 1); // offset is 1-based, location is 0-based
+
+        for (const { fileName, textChanges } of edits) {
+            if (fileName !== locationFileName) {
+                continue;
+            }
+
+            for (const { newText, span: { start, length } } of textChanges) {
+                const newTextNewlines = countNewlines(newText);
+                line += newTextNewlines - countNewlines(oldText.slice(start, start + length));
+                if (start > lineStartOfLocation && start < location) {
+                    offset += newText.length - length;
+                    if (newTextNewlines !== 0) {
+                        offset -= newText.lastIndexOf("\n");
+                    }
+                }
+            }
+        }
+
+        return { line, offset };
+    }
+
+    function countNewlines(s: string): number {
+        let count = 0;
+        for (let i = 0; i < s.length; i++) {
+            if (s.charCodeAt(i) === CharacterCodes.lineFeed) {

```

Comment:
It looks like this will do the wrong thing for CR-only files.  I assume we don't care.